sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala

```diff
@@ -36,107 +35,12 @@ import org.apache.spark.sql.internal.SQLConf
-            e.copy(coordinator = Some(coordinator))
-          case (child, distribution) =>
-            // If this child is not an Exchange, we need to add an Exchange for now.
-            // Ideally, we can try to avoid this Exchange. However, when we reach here,
-            // there are at least two children operators (because if there is a single child
-            // and we can avoid Exchange, supportsCoordinator will be false and we
-            // will not reach here.). Although we can make two children have the same number of
-            // post-shuffle partitions. Their numbers of pre-shuffle partitions may be different.
-            // For example, let's say we have the following plan
-            //         Join
-            //         /  \
-            //       Agg  Exchange
-            //       /      \
-            //    Exchange  t2
-            //      /
-            //     t1
-            // In this case, because a post-shuffle partition can include multiple pre-shuffle
-            // partitions, a HashPartitioning will not be strictly partitioned by the hashcodes
-            // after shuffle. So, even we can use the child Exchange operator of the Join to
-            // have a number of post-shuffle partitions that matches the number of partitions of
-            // Agg, we cannot say these two children are partitioned in the same way.
-            // Here is another case
-            //         Join
-            //         /  \
-            //       Agg1  Agg2
-            //       /      \
-            //   Exchange1  Exchange2
-            //       /       \
-            //      t1       t2
-            // In this case, two Aggs shuffle data with the same column of the join condition.
-            // After we use ExchangeCoordinator, these two Aggs may not be partitioned in the same
-            // way. Let's say that Agg1 and Agg2 both have 5 pre-shuffle partitions and 2
-            // post-shuffle partitions. It is possible that Agg1 fetches those pre-shuffle
-            // partitions by using a partitionStartIndices [0, 3]. However, Agg2 may fetch its
-            // pre-shuffle partitions by using another partitionStartIndices [0, 4].
-            // So, Agg1 and Agg2 are actually not co-partitioned.
-            //
-            // It will be great to introduce a new Partitioning to represent the post-shuffle
-            // partitions when one post-shuffle partition includes multiple pre-shuffle partitions.
-            val targetPartitioning = distribution.createPartitioning(defaultNumPreShufflePartitions)
-            assert(targetPartitioning.isInstanceOf[HashPartitioning])
-            ShuffleExchangeExec(targetPartitioning, child, Some(coordinator))
-        }
-      } else {
-        // If we do not need ExchangeCoordinator, the original children are returned.
-        children
-      }
-
-    withCoordinator
-  }
+  private def defaultNumPreShufflePartitions: Int =
```

Comment:
@carsonwang  With AE being the new mode, spark.sql.adaptive.maxNumPostShufflePartitions will replace spark.sql.shuffle.partitions to determine initial shuffle parallelism, this seems to be a significant user-facing change. E.g., if a user has set spark.sql.shuffle.partitions to 10K for a large join, with AE enabled he has to set maxNumPostShufflePartitions to 10K otherwise he will only get 500. I made a small patch when testing AE for production jobs, by setting maxNumPostShufflePartitions = spark.sql.shuffle.partitions, if spark.sql.shuffle.partitions != default/200. What do you think？

Reply:
If the user is familiar with the AE configuration, can he just set the maxNumPostShufflePartitions for the large join when he enables AE? If he set spark.sql.shuffle.partitions to 10K in non-AE mode. I expect he will set a higher value for maxNumPostShufflePartitions like 20K in AE mode. So AE can find a good reducer number between 1 and 20K and usually a better value than 10K.

Reply:
@carsonwang, I am afraid we are making a risky assumption that user needs to be familiar with AE config. it is less risky when AE is an on-demand feature. however, I find the current version of this PR is setting spark.sql.adaptive.enabled = true, which means we plan to enable AE mode by default, then if we roll out next version of Spark in our cluster we are going to break a lot of prod jobs with custom spark.sql.shuffle.partitions.  I proposed a small change to adjust maxNumPostShufflePartitions based on spark.sql.shuffle.partitions which I think is safer, if the plan is to set AE as cluster default in the future

Reply:
```
 I proposed a change to adjust maxNumPostShufflePartitions based on spark.sql.shuffle.partitions which I think is safer.
```
I think it makes sense. Actually, during our internal practice, we set the default value of maxNumPostShufflePartitions = 1.5 * spark.sql.shuffle.partitions. But for the common code here, a magic number of 1.5 here cause doubt, maybe we need a discussion for the strategy of setting an appropriate default value for `maxNumPostShufflePartitions` based on `spark.sql.shuffle.partitions`.

Reply:
Thanks @xuanyuanking for the input! Instead of setting maxNumPostShufflePartitions based on a magic number 1.5X or 500, I would propose to add a conf to replace maxNumPostShufflePartitions, which is a ratio of spark.sql.shuffle.partitions. The default value could be 1.0 so the behavior of AE's initial partition number is consistent with user-specified spark.sql.shuffle.partitions. With 1.5 or 2, one of my concern is it could potentially be a sharp increase in shuffle service load at the time we enable AE as cluster default, and we have seen shuffle service scalability issues in our cluster when handling large shuffle workloads. 
cc @carsonwang  @cloud-fan

Reply:
It is a good point we should not break anything when AE is enabled by default at cluster level. Currently it is enabled only for test purpose. But it is possible we enable it by default in future. What about we set maxNumPostShufflePartitions to spark.sql.shuffle.partitions by default? We are currently using the maxNumPostShufflePartitions as an initial partition number. But in future, if we can find a better initial partition number between minNumPostShufflePartitions and maxNumPostShufflePartitions at runtime, maxNumPostShufflePartitions will be treated as a max limit.

Reply:
@carsonwang, great, I support setting maxNumPostShufflePartitions = spark.sql.shuffle.partitions by default. Based on @xuanyuanking 's scenario, setting maxNumPostShufflePartitions as a configurable ratio to spark.sql.shuffle.partitions would also make sense to me.